cmd/tusd/cli/hooks/grpc.go

```diff
@@ -0,0 +1,111 @@
+package hooks
+
+import (
+	"context"
+	"io"
+	"log"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	grpc_retry "github.com/grpc-ecosystem/go-grpc-middleware/retry"
+	"github.com/tus/tusd/pkg/handler"
+	pb "github.com/tus/tusd/pkg/proto/v1"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/status"
+)
+
+type GrpcHook struct {
+	Endpoint   string
+	MaxRetries int
+	Backoff    int
+}
+
+func (_ GrpcHook) Setup() error {
+	return nil
+}
+
+func (g GrpcHook) InvokeHook(typ HookType, info handler.HookEvent, captureOutput bool) ([]byte, int, error) {
+	ctx := context.Background()
+	opts := []grpc_retry.CallOption{
+		grpc_retry.WithBackoff(grpc_retry.BackoffLinear(time.Duration(g.Backoff) * time.Second)),
+		grpc_retry.WithMax(uint(g.MaxRetries)),
+	}
+	conn, err := grpc.Dial(g.Endpoint,
+		grpc.WithInsecure(),
+		grpc.WithStreamInterceptor(grpc_retry.StreamClientInterceptor(opts...)),
+		grpc.WithUnaryInterceptor(grpc_retry.UnaryClientInterceptor(opts...)),
+	)
+	if err != nil {
+		return nil, 0, err
+	}
+	defer conn.Close()
+	client := pb.NewHookServiceClient(conn)
+	stream, err := client.Send(ctx)
+	if err != nil {
+		if e, ok := status.FromError(err); ok {
+			return nil, int(e.Code()), err
+		}
+		return nil, 0, err
+	}
+
+	var data []byte
+	var resp *pb.SendResponse
+	ctx = stream.Context()
+	done := make(chan bool)
+
+	req := &pb.SendRequest{Hook: marshal(info)}
+	if err := stream.Send(req); err != nil {
+		log.Fatalf("can not send %v", err)
+	}
+	if err := stream.CloseSend(); err != nil {
+		log.Println(err)
+	}
+	go func() {
```

Comment:
Can you explain why goroutines are necessary here?

Reply:
> When using streams, one must take care to avoid calling either SendMsg or RecvMsg multiple times against the same Stream from different goroutines. In other words, it's safe to have a goroutine calling SendMsg and another goroutine calling RecvMsg on the same stream at the same time. 
[link](https://github.com/grpc/grpc-go/blob/master/Documentation/concurrency.md#streams)